compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java

```diff
@@ -757,7 +757,7 @@ private void generateFrameClassFieldLoad(List<BIRVariableDcl> localVars, MethodV
             } else if (bType.tag == TypeTags.NEVER) {
                 mv.visitFieldInsn(GETFIELD, frameName, localVar.jvmVarName, GET_OBJECT);
                 mv.visitVarInsn(ASTORE, index);
-            } else if (types.isAssignable(bType, symbolTable.xmlType)) {
+            } else if (TypeTags.isXMLTypeTag(bType.tag)) {
```

Comment:
By doing this, if the `btype` is a union type, we cannot identify whether it is an XML type or not.
As I understand, the reason to do a cast in desugar is to avoid btype becoming a union. If that's the case, I think that cast is unnecessary. Can't we fix this issue here for union types instead of relying on the cast in desugar?

Reply:
In the runtime we represent union types using the OBJECT type in the frame class. Thus for an xml union type it is wrong here to go inside this if condition. This was an invalid fix introduced here. https://github.com/ballerina-platform/ballerina-lang/pull/42150/files.

Reply:
My suggestion is that we do not cast here. Instead, in the `MethodGen`, if the type is a union and a subtype of XML, we can update the logic accordingly.

My concern with creating the cast here is that it is not included in the generated bytecode. Isn't that correct?

Reply:
Yes, for union types, we need to handle them separately.